base/task.jl

```diff
@@ -365,6 +365,142 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+
+"""
+    waitany(tasks; throw=false) -> (done_tasks, remaining_tasks)
+
+Wait until at least one of the given tasks have been completed.
+
+If `throw` is `true`, throw `CompositeException` when one of the
+completed tasks completes with an exception.
+
+The return value consists of two task vectors. The first one consists of
+completed tasks, and the other consists of uncompleted tasks.
+"""
+waitany(tasks; throw=false) = _wait_multiple(tasks, throw)
+
+"""
+    waitall(tasks; failfast=false, throw=false) -> (done_tasks, remaining_tasks)
+
+Wait until all the given tasks have been completed.
+
+If `failfast` is `true`, the function will return when at least one of the
+given tasks is finished by exception. If `throw` is `true`, throw
+`CompositeException` when one of the completed tasks has failed.
+
+`failfast` and `throw` keyword arguments work independently; when only
+`throw=true` is specified, this function waits for all the tasks to complete.
+
+The return value consists of two task vectors. The first one consists of
+completed tasks, and the other consists of uncompleted tasks.
+"""
+waitall(tasks; failfast=false, throw=false) = _wait_multiple(tasks, throw, true, failfast)
+
+function _wait_multiple(waiting_tasks, throwexc=false, all=false, failfast=false)
+    tasks = Task[]
+
+    for t in waiting_tasks
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+    end
+
```

Comment:
@vtjnash How do you think split `_wait_multiple` for `AbstractVector{Task}` and other types?

```suggestion
function _wait_multiple(waiting_tasks, throwexc=false, all=false, failfast=false)
    tasks = Task[]

    for t in waiting_tasks
        t isa Task || error("Expected an iterator of `Task` object")
        push!(tasks, t)
    end
    
    return _wait_multiple(tasks, throwexc, all, failfast)
end

function _wait_multiple(tasks::AbstractVector{Task}, throwexc=false, all=false, failfast=false)
```

Reply:
To avoid the copy? That makes sense to me. Though I don't know if all AbstractVector support the bitmasking operations, I think that should be valid.

As food for thought also, in a later PR, I am thinking we should also seek to extend this to allow any other waitable events too, not just Tasks (e.g. Channels, Event, AsyncEvent, Timer, etc.) and either add the `throwexc` kwarg to those other `wait` functions (instead of calling it `_wait`) or rename `_wait` to something like `waitready`

Reply:
> Though I don't know if all AbstractVector support the bitmasking operations, I think that should be valid.

I guess `a[mask]` is processed by `_getindex(IndexStyle(a), a, LogicalIndex(mask)...)`, which then reaches `_unsafe_getindex` defined in multidimensional.jl, so all AbstractVector support the bitmasking operations.

> As food for thought also, in a later PR, I am thinking we should also seek to extend this to allow any other waitable events too, not just Tasks

I completely agree. As I wrote in issue #53226, I also thought that supporting waitable objects other than Tasks might be a good idea.

> either add the `throwexc` kwarg to those other `wait` functions (instead of calling it `_wait`) or rename `_wait` to something like `waitready`

I think the non-throw version might be used as frequently as `wait` in real-world application development.

Reply:
I submitted a new PR #53685 for introducing `throw` option in `wait(::Task)` function.